File: gitgossipfrontend/lib/features/chat/services/socket_service.dart
import 'package:socket_io_client/socket_io_client.dart' as IO;

// socket_service → manages socket connection
class SocketService {
  static final SocketService _instance = SocketService._internal();

  factory SocketService() {
    return _instance;
  }

  SocketService._internal();

  IO.Socket? socket;

  void connect() {
    if (socket != null && socket!.connected) {
      return;
    }
    socket = IO.io(
      "http://10.148.217.72:3000",
      IO.OptionBuilder()
          .setTransports(['websocket'])
          .disableAutoConnect()
          .build(),
    );

    socket!.connect();
    listenEvents();
  }

  // Listen Core Socket Events
  void listenEvents() {
    socket!.onConnect((_) {
      print("✅ Socket Connected: ${socket!.id}");
    });

    socket!.onDisconnect((_) {
      print("❌ Socket Disconnected");
    });

    socket!.onConnectError((error) {
      print("⚠️ Connection Error: $error");
    });

    socket!.onError((error) {
      print("⚠️ Socket Error: $error");
    });
  }

  // Emit Events //
  // Join Chat Room
  void joinConversation(String conversationId) {
    socket?.emit("joinConversation", conversationId);
  }

  // Send Message
  void sendMessage(Map<String, dynamic> data) {
    socket?.emit("sendMessage", data);
  }

  // Listen Message
  void onReceiveMessage(Function(dynamic) callback) {
    socket?.on("receiveMessage", callback);
  }

  // Message Seen
  void onMessageSeen(Function(dynamic) callback) {
    socket?.on("messagesSeen", callback);
  }
}

File: errorbackend/src/services/socketService.js
// handle connection
// listen sendMessage
// emit receiveMessage

import Conversation from "../models/conversationModel.js";
import Message from "../models/messageModel.js";


export const initSocket = (io) => {

    io.on("connection", (socket) => {

        console.log("User connected:", socket.id);

        // join conversation room
        socket.on("joinConversation", (conversationId) => {
            socket.join(conversationId);
        });

        // send message
        socket.on("sendMessage", async (data) => {

            const { conversationId, senderId, text } = data;

            const message = await Message.create({
                conversationId,
                senderId,
                text,
            });
            await Conversation.findByIdAndUpdate(conversationId, {
                lastMessage: message._id,
                lastMessageSenderId: senderId,
            })
            // emit message to all users in that conversation
            io.to(conversationId).emit("receiveMessage", message);
        });

        socket.on("disconnect", () => {
            console.log("User disconnected:", socket.id);
        });

        socket.on("markSeen", async (data) => {

            const { conversationId, userId } = data;

            await Message.updateMany(
                {
                    conversationId,
                    senderId: { $ne: userId },
                    seen: false
                },
                { $set: { seen: true } }
            );

            io.to(conversationId).emit("messagesSeen", {
                conversationId
            });

        });


    });

};